File: apps/sandbox-runtime/internal/tunnel/messages.go
package tunnel

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/coder/websocket"
	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/sessionprotocol"
)

const (
	connectErrorCodeUnsupportedChannel         = "unsupported_channel"
	connectErrorCodeInvalidConnectRequest      = "invalid_connect_request"
	connectErrorCodeAgentEndpointUnavailable   = "agent_endpoint_unavailable"
	connectErrorCodeUnsupportedConnectionMode  = "unsupported_connection_mode"
	connectErrorCodeAgentEndpointDialFailed    = "agent_endpoint_dial_failed"
	connectErrorCodePTYSessionUnavailable      = "pty_session_unavailable"
	connectErrorCodePTYSessionExists           = "pty_session_exists"
	connectErrorCodePTYSessionCreateFailed     = "pty_session_create_failed"
	ptyConnectErrorCodeInvalidSessionSelection = "invalid_pty_session_mode"
	streamResetCodeInvalidStreamSignal         = "invalid_stream_signal"
	streamResetCodeInvalidStreamClose          = "invalid_stream_close"
	streamResetCodeInvalidStreamData           = "invalid_stream_data"
	streamResetCodeInvalidStreamWindow         = "invalid_stream_window"
	streamResetCodeStreamCloseFailed           = "stream_close_failed"
	streamResetCodeTargetClosed                = "target_closed"
)

func writeStreamOpenOK(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamOpenOK sessionprotocol.StreamOpenOK,
) error {
	return writeTextJSONMessage(ctx, tunnelConn, streamOpenOK)
}

func writeStreamOpenError(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamOpenError sessionprotocol.StreamOpenError,
) error {
	return writeTextJSONMessage(ctx, tunnelConn, streamOpenError)
}

func writeStreamEvent(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamEvent sessionprotocol.StreamEvent,
) error {
	return writeTextJSONMessage(ctx, tunnelConn, streamEvent)
}

func writeStreamReset(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamReset sessionprotocol.StreamReset,
) error {
	return writeTextJSONMessage(ctx, tunnelConn, streamReset)
}

func writeStreamWindow(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamWindow sessionprotocol.StreamWindow,
) error {
	return writeTextJSONMessage(ctx, tunnelConn, streamWindow)
}

func writeTextJSONMessage(ctx context.Context, tunnelConn *websocket.Conn, payload any) error {
	encodedPayload, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("failed to encode json payload: %w", err)
	}

	if err := tunnelConn.Write(ctx, websocket.MessageText, encodedPayload); err != nil {
		return fmt.Errorf("failed to write websocket message: %w", err)
	}

	return nil
}

func writeBinaryDataFrame(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	streamID uint32,
	payloadKind byte,
	payload []byte,
) error {
	encodedPayload, err := sessionprotocol.EncodeDataFrame(struct {
		StreamID    uint32
		PayloadKind byte
		Payload     []byte
	}{
		StreamID:    streamID,
		PayloadKind: payloadKind,
		Payload:     payload,
	})
	if err != nil {
		return fmt.Errorf("failed to encode data frame: %w", err)
	}

	if err := tunnelConn.Write(ctx, websocket.MessageBinary, encodedPayload); err != nil {
		return fmt.Errorf("failed to write websocket message: %w", err)
	}

	return nil
}

File: apps/sandbox-runtime-node/src/tunnel/messages.ts
import {
  encodeDataFrame,
  type ExecutionLease,
  type LeaseCreate,
  type LeaseRenew,
  type StreamClose,
  type StreamEventMessage,
  type StreamOpenError,
  type StreamOpenOK,
  type StreamReset,
  type StreamWindow,
} from "@mistle/sandbox-session-protocol";
import type WebSocket from "ws";

export const CONNECT_ERROR_CODE_UNSUPPORTED_CHANNEL = "unsupported_channel";
export const CONNECT_ERROR_CODE_INVALID_CONNECT_REQUEST = "invalid_connect_request";
export const CONNECT_ERROR_CODE_AGENT_ENDPOINT_UNAVAILABLE = "agent_endpoint_unavailable";
export const CONNECT_ERROR_CODE_UNSUPPORTED_CONNECTION_MODE = "unsupported_connection_mode";
export const CONNECT_ERROR_CODE_AGENT_ENDPOINT_DIAL_FAILED = "agent_endpoint_dial_failed";
export const CONNECT_ERROR_CODE_PTY_SESSION_UNAVAILABLE = "pty_session_unavailable";
export const CONNECT_ERROR_CODE_PTY_SESSION_EXISTS = "pty_session_exists";
export const CONNECT_ERROR_CODE_PTY_SESSION_CREATE_FAILED = "pty_session_create_failed";
export const PTY_CONNECT_ERROR_CODE_INVALID_SESSION_SELECTION = "invalid_pty_session_mode";
export const STREAM_RESET_CODE_INVALID_STREAM_SIGNAL = "invalid_stream_signal";
export const STREAM_RESET_CODE_INVALID_STREAM_CLOSE = "invalid_stream_close";
export const STREAM_RESET_CODE_INVALID_STREAM_DATA = "invalid_stream_data";
export const STREAM_RESET_CODE_INVALID_STREAM_WINDOW = "invalid_stream_window";
export const STREAM_RESET_CODE_STREAM_CLOSE_FAILED = "stream_close_failed";
export const STREAM_RESET_CODE_TARGET_CLOSED = "target_closed";
export const STREAM_RESET_CODE_STREAM_WINDOW_EXHAUSTED = "stream_window_exhausted";

function sendWebSocketPayload(
  socket: WebSocket,
  payload: string | Uint8Array,
  isBinary: boolean,
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    socket.send(payload, { binary: isBinary }, (error: Error | undefined) => {
      if (error === undefined) {
        resolve();
        return;
      }

      reject(error);
    });
  });
}

async function writeTextJsonMessage(socket: WebSocket, payload: object): Promise<void> {
  await sendWebSocketPayload(socket, JSON.stringify(payload), false);
}

export function writeStreamOpenOk(socket: WebSocket, streamOpenOk: StreamOpenOK): Promise<void> {
  return writeTextJsonMessage(socket, streamOpenOk);
}

export function writeStreamOpenError(
  socket: WebSocket,
  streamOpenError: StreamOpenError,
): Promise<void> {
  return writeTextJsonMessage(socket, streamOpenError);
}

export function writeStreamEvent(
  socket: WebSocket,
  streamEvent: StreamEventMessage,
): Promise<void> {
  return writeTextJsonMessage(socket, streamEvent);
}

export function writeStreamReset(socket: WebSocket, streamReset: StreamReset): Promise<void> {
  return writeTextJsonMessage(socket, streamReset);
}

export function writeStreamWindow(socket: WebSocket, streamWindow: StreamWindow): Promise<void> {
  return writeTextJsonMessage(socket, streamWindow);
}

export function writeLeaseCreate(socket: WebSocket, lease: ExecutionLease): Promise<void> {
  const message: LeaseCreate = {
    type: "lease.create",
    lease,
  };

  return writeTextJsonMessage(socket, message);
}

export function writeLeaseRenew(socket: WebSocket, leaseId: string): Promise<void> {
  const message: LeaseRenew = {
    type: "lease.renew",
    leaseId,
  };

  return writeTextJsonMessage(socket, message);
}

export function writeBinaryDataFrame(
  socket: WebSocket,
  input: {
    streamId: number;
    payloadKind: number;
    payload: Uint8Array;
  },
): Promise<void> {
  return sendWebSocketPayload(
    socket,
    encodeDataFrame({
      streamId: input.streamId,
      payloadKind: input.payloadKind,
      payload: input.payload,
    }),
    true,
  );
}

export function createPtyExitEventMessage(streamId: number, exitCode: number): StreamEventMessage {
  return {
    type: "stream.event",
    streamId,
    event: {
      type: "pty.exit",
      exitCode,
    },
  };
}

export function createStreamClose(streamId: number): StreamClose {
  return {
    type: "stream.close",
    streamId,
  };
}

export function createPtyResizeSignal(
  streamId: number,
  cols: number,
  rows: number,
): {
  type: "stream.signal";
  streamId: number;
  signal: {
    type: "pty.resize";
    cols: number;
    rows: number;
  };
} {
  return {
    type: "stream.signal",
    streamId,
    signal: {
      type: "pty.resize",
      cols,
      rows,
    },
  };
}
